src/items.rs

```diff
@@ -135,22 +139,42 @@ pub enum Formula {
     Eq(FormulaList),
 }
 
-impl ToString for Formula {
-    fn to_string(&self) -> String {
-        let fl_to_string = |fl: &FormulaList| {
-            let fv: Vec<String> = fl.iter().map(|x| x.to_string()).collect();
-            fv.join(" ")
+impl Display for Formula {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        // Prefix
+        match self {
+            Formula::Lit(l) => write!(f, "{}", l)?,
+            Formula::Paren(m) => write!(f, "({})", m)?,
+            Formula::Neg(m) => write!(f, "-{}", m)?,
+            Formula::And(_) => write!(f, "*(")?,
+            Formula::Or(_) => write!(f, "+(")?,
+            Formula::Xor(_) => write!(f, "xor(")?,
+            Formula::Eq(_) => write!(f, "=(")?,
         };
+
+        // Suffix
         match self {
-            Formula::Lit(l) => l.to_string(),
-            Formula::Paren(f) => format!("({})", f.to_string()),
-            Formula::Neg(f) => format!("-{}", f.to_string()),
-            Formula::And(fl) => format!("*({})", fl_to_string(fl)),
-            Formula::Or(fl) => format!("+({})", fl_to_string(fl)),
-            Formula::Xor(fl) => format!("xor({})", fl_to_string(fl)),
-            Formula::Eq(fl) => format!("=({})", fl_to_string(fl)),
+            Formula::Lit(_) | Formula::Paren(_) | Formula::Neg(_) => {}
+            Formula::And(m)
+            | Formula::Or(m)
+            | Formula::Xor(m)
+            | Formula::Eq(m) => {
+                for idx in 0..(m.len() - 1) {
+                    write!(f, "{} ", m[idx])?;
+                }
+                if m.len() != 0 {
```

Comment:
prefer using `m.is_empty()`

Reply:
besides that I'd rather like to see a solution using Rust slice's `split_first()` for this whole chunk.